File: packages/crypto/src/nullifier.ts
/**
 * @file nullifier.ts
 * @description Client-side Tier 1 voting crypto for ekklesia.gr.
 *
 * Protocol summary (Tier 1 — HMAC-Nullifier-Chain):
 *   1. Registration: Argon2id(phone) → nullifier_root → identity_commitment → server
 *   2. Voting:       nullifier_root → ephemeral Ed25519 keypair (per bill)
 *                                   → vote_nullifier (per bill, unique, not linkable)
 *                                   → linkage_tag   (proves shared root, trust-based)
 *                    Sign payload with ephemeral key → send → discard key
 *
 * Trust model (Tier 1):
 *   Server cannot cryptographically verify key derivation without ZK.
 *   Guarantee: open-source client + Argon2id cost makes Sybil expensive.
 *   Full cryptographic guarantee: Tier 2 (Semaphore/Groth16).
 *
 * @module @ekklesia/crypto/nullifier
 */

import { hmac }           from "@noble/hashes/hmac";
import { sha256 }         from "@noble/hashes/sha256";
import { ed25519 }        from "@noble/curves/ed25519";
import { argon2id }       from "hash-wasm";
import {
  bytesToHex,
  hexToBytes,
  utf8ToBytes,
  concatBytes,
}                          from "@noble/hashes/utils";
import {
  ARGON2_PARAMS,
  DOMAIN,
  PROTO_VERSION,
  REGISTRATION_SALT,
  type PhoneRegion,
  type RegistrationPayload,
  type VoteChoice,
  type VotePayload,
  type VoteReceipt,
} from "./types.js";

// ─── Internal helpers ─────────────────────────────────────────────────────────

/**
 * HMAC-SHA256 over UTF-8 domain string + optional binary suffix.
 * Always returns 32 bytes.
 */
function hmacSha256(key: Uint8Array, domain: string, suffix?: Uint8Array): Uint8Array {
  const domainBytes = utf8ToBytes(domain);
  const message     = suffix ? concatBytes(domainBytes, suffix) : domainBytes;
  return hmac(sha256, key, message);
}

/**
 * Normalize Greek phone number to E.164.
 * Accepts: 6912345678 / 06912345678 / +306912345678 / 00306912345678
 */
function normalizePhone(raw: string): string {
  const digits = raw.replace(/\D/g, "");
  if (digits.startsWith("30") && digits.length === 12) return `+${digits}`;
  if (digits.startsWith("0030"))                        return `+30${digits.slice(4)}`;
  if (digits.length === 10 && digits.startsWith("6"))   return `+30${digits}`;
  throw new Error(`Cannot normalize phone number: ${raw}`);
}

/**
 * Zero out a Uint8Array in-place (best-effort in JS — GC may copy).
 * Do not rely on this for hard security guarantees.
 */
function zero(buf: Uint8Array): void {
  buf.fill(0);
}

// ─── Storage helpers (IndexedDB) ──────────────────────────────────────────────

const DB_NAME    = "ekklesia_crypto";
const DB_VERSION = 1;
const STORE_NAME = "identity";

function openDb(): Promise<IDBDatabase> {
  return new Promise((resolve, reject) => {
    const req = indexedDB.open(DB_NAME, DB_VERSION);
    req.onupgradeneeded = () => req.result.createObjectStore(STORE_NAME);
    req.onsuccess       = () => resolve(req.result);
    req.onerror         = () => reject(req.error);
  });
}

async function dbGet(key: string): Promise<Uint8Array | undefined> {
  const db  = await openDb();
  const tx  = db.transaction(STORE_NAME, "readonly");
  const req = tx.objectStore(STORE_NAME).get(key);
  return new Promise((resolve, reject) => {
    req.onsuccess = () => resolve(req.result as Uint8Array | undefined);
    req.onerror   = () => reject(req.error);
  });
}

async function dbSet(key: string, value: Uint8Array): Promise<void> {
  const db  = await openDb();
  const tx  = db.transaction(STORE_NAME, "readwrite");
  tx.objectStore(STORE_NAME).put(value, key);
  return new Promise((resolve, reject) => {
    tx.oncomplete = () => resolve();
    tx.onerror    = () => reject(tx.error);
  });
}

async function dbDelete(key: string): Promise<void> {
  const db  = await openDb();
  const tx  = db.transaction(STORE_NAME, "readwrite");
  tx.objectStore(STORE_NAME).delete(key);
  return new Promise((resolve, reject) => {
    tx.oncomplete = () => resolve();
    tx.onerror    = () => reject(tx.error);
  });
}

// ─── Public API ───────────────────────────────────────────────────────────────

/**
 * Derives the nullifier_root from a phone number using Argon2id.
 *
 * This is the root secret from which all voting keys are derived.
 * NEVER send this to the server or log it.
 *
 * @param phone - Raw phone number (any common format accepted)
 * @returns 32-byte nullifier_root
 */
export async function deriveNullifierRoot(phone: string): Promise<Uint8Array> {
  const normalized = normalizePhone(phone);
  const result     = await argon2id({
    password:    normalized,
    salt:        REGISTRATION_SALT,
    ...ARGON2_PARAMS,
  });
  // hash-wasm returns hex string when outputType is not set to "binary" in some versions
  if (typeof result === "string") return hexToBytes(result);
  return result as Uint8Array;
}

/**
 * Persists nullifier_root to IndexedDB under key "nullifier_root".
 *
 * Security note: IndexedDB is not hardware-backed.
 * XSS on the domain can read it. Mitigate with strict CSP.
 */
export async function persistNullifierRoot(nullifierRoot: Uint8Array): Promise<void> {
  await dbSet("nullifier_root", nullifierRoot);
}

/**
 * Loads nullifier_root from IndexedDB.
 * Returns undefined if not registered on this device.
 */
export async function loadNullifierRoot(): Promise<Uint8Array | undefined> {
  return dbGet("nullifier_root");
}

/**
 * Checks whether this device has a registered identity.
 */
export async function isRegistered(): Promise<boolean> {
  const root = await loadNullifierRoot();
  return root !== undefined;
}

/**
 * Derives the identity_commitment from nullifier_root.
 * This is the ONLY value sent to the server during registration.
 * It cannot be used to reconstruct nullifier_root.
 *
 * @param nullifierRoot - 32-byte root secret
 * @returns identity_commitment as hex string
 */
export function deriveIdentityCommitment(nullifierRoot: Uint8Array): string {
  const commitment = hmacSha256(nullifierRoot, DOMAIN.IDENTITY_COMMITMENT);
  return bytesToHex(commitment);
}

/**
 * Builds the registration payload to send to the server.
 *
 * @param phone       - Raw phone number
 * @param phoneRegion - ISO 3166-1 alpha-2 (e.g. "GR")
 * @param persist     - Whether to persist nullifier_root to IndexedDB (default: true)
 * @returns { payload, nullifierRoot } — caller must handle nullifierRoot carefully
 */
export async function buildRegistrationPayload(
  phone: string,
  phoneRegion: PhoneRegion,
  persist = true,
): Promise<{ payload: RegistrationPayload; nullifierRoot: Uint8Array }> {
  const nullifierRoot         = await deriveNullifierRoot(phone);
  const identity_commitment   = deriveIdentityCommitment(nullifierRoot);

  if (persist) await persistNullifierRoot(nullifierRoot);

  const payload: RegistrationPayload = {
    identity_commitment,
    phone_region: phoneRegion,
    version:      PROTO_VERSION,
  };

  return { payload, nullifierRoot };
}

// ─── Voting ───────────────────────────────────────────────────────────────────

/**
 * Derives the vote_nullifier for a specific bill.
 * Unique per (user × bill). Not linkable across bills without nullifier_root.
 *
 * @param nullifierRoot - 32-byte root secret
 * @param billId        - Bill identifier e.g. "GR-2026-0042"
 * @returns vote_nullifier as hex string
 */
export function deriveVoteNullifier(nullifierRoot: Uint8Array, billId: string): string {
  const nullifier = hmacSha256(
    nullifierRoot,
    DOMAIN.VOTE_NULLIFIER + ":" + billId,
  );
  return bytesToHex(nullifier);
}

/**
 * Derives the linkage_tag for a specific bill.
 *
 * Purpose: allows server to verify (in a trust-based way) that
 * the vote_nullifier and the identity_commitment share the same nullifier_root.
 * In Tier 2 this is replaced by a ZK-proof.
 *
 * @param nullifierRoot - 32-byte root secret
 * @param billId        - Bill identifier
 * @returns linkage_tag as hex string
 */
export function deriveLinkageTag(nullifierRoot: Uint8Array, billId: string): string {
  const linkageKey = hmacSha256(nullifierRoot, DOMAIN.LINKAGE_TAG);
  const tag        = hmacSha256(linkageKey, billId);
  return bytesToHex(tag);
}

/**
 * Derives a deterministic ephemeral Ed25519 keypair for a specific bill.
 * Deterministic: same (nullifier_root × bill_id) always yields the same keypair.
 * Ephemeral: private key must be zeroed after signing.
 *
 * @param nullifierRoot - 32-byte root secret
 * @param billId        - Bill identifier
 * @returns { sk, pk } — caller MUST call zero(sk) after use
 */
export function deriveEphemeralKeypair(
  nullifierRoot: Uint8Array,
  billId: string,
): { sk: Uint8Array; pk: Uint8Array } {
  const seed = hmacSha256(
    nullifierRoot,
    DOMAIN.EPHEMERAL_SEED + ":" + billId,
  );
  const sk = seed;                       // 32-byte private key seed
  const pk = ed25519.getPublicKey(sk);   // derive public key
  return { sk, pk };
}

/**
 * Builds the canonical byte payload that is signed by the ephemeral key.
 * All fields are length-prefixed to prevent collision attacks.
 *
 * Layout: version(1B) | bill_id_len(2B) | bill_id | choice(1B) |
 *         pk_eph(32B) | vote_nullifier(32B) | linkage_tag(32B) | timestamp(8B)
 */
export function buildSignedPayload(
  billId:        string,
  choice:        VoteChoice,
  pkEph:         Uint8Array,
  voteNullifier: Uint8Array,
  linkageTag:    Uint8Array,
  timestampMs:   number,
): Uint8Array {
  const versionByte  = new Uint8Array([1]);
  const billBytes    = utf8ToBytes(billId);
  const billLenBytes = new Uint8Array(2);
  new DataView(billLenBytes.buffer).setUint16(0, billBytes.length, false);

  const choiceMap: Record<VoteChoice, number> = { YES: 1, NO: 2, ABSTAIN: 3 };
  const choiceByte = new Uint8Array([choiceMap[choice]]);

  const tsBuf = new ArrayBuffer(8);
  new DataView(tsBuf).setBigUint64(0, BigInt(timestampMs), false);
  const tsBytes = new Uint8Array(tsBuf);

  return concatBytes(
    versionByte, billLenBytes, billBytes,
    choiceByte, pkEph, voteNullifier, linkageTag, tsBytes,
  );
}

/**
 * Builds and signs a vote payload.
 * Ephemeral private key is zeroed immediately after signing.
 *
 * @param nullifierRoot - 32-byte root secret (from loadNullifierRoot())
 * @param billId        - Bill identifier
 * @param choice        - Vote choice
 * @returns Signed VotePayload ready to POST to /api/votes
 */
export function buildVotePayload(
  nullifierRoot: Uint8Array,
  billId:        string,
  choice:        VoteChoice,
): VotePayload {
  const timestampMs   = Date.now();
  const { sk, pk }    = deriveEphemeralKeypair(nullifierRoot, billId);
  const voteNullifier = hexToBytes(deriveVoteNullifier(nullifierRoot, billId));
  const linkageTag    = hexToBytes(deriveLinkageTag(nullifierRoot, billId));

  const payloadBytes = buildSignedPayload(
    billId, choice, pk, voteNullifier, linkageTag, timestampMs,
  );
  const signature = ed25519.sign(payloadBytes, sk);

  // Zero the ephemeral private key immediately (best-effort)
  zero(sk);

  return {
    bill_id:        billId,
    choice,
    pk_eph:         bytesToHex(pk),
    vote_nullifier: bytesToHex(voteNullifier),
    linkage_tag:    bytesToHex(linkageTag),
    signature:      bytesToHex(signature),
    timestamp_ms:   timestampMs,
    version:        PROTO_VERSION,
  };
}

// ─── Receipt verification ─────────────────────────────────────────────────────

/**
 * Verifies a server-issued VoteReceipt.
 * Clients SHOULD call this after receiving a successful vote response.
 *
 * @param receipt      - VoteReceipt from server
 * @param voteNullifier - The vote_nullifier we sent (to verify it matches)
 * @returns true if receipt is genuine and matches our vote
 */
export function verifyReceipt(
  receipt:        VoteReceipt,
  voteNullifier:  string,
): boolean {
  if (receipt.vote_nullifier !== voteNullifier) return false;

  const receiptBytes = concatBytes(
    utf8ToBytes(receipt.bill_id),
    hexToBytes(receipt.vote_nullifier),
    (() => {
      const buf = new ArrayBuffer(8);
      new DataView(buf).setBigUint64(0, BigInt(receipt.server_timestamp_ms), false);
      return new Uint8Array(buf);
    })(),
  );

  try {
    return ed25519.verify(
      hexToBytes(receipt.server_signature),
      receiptBytes,
      hexToBytes(receipt.server_pk),
    );
  } catch {
    return false;
  }
}

// ─── Receipt storage ──────────────────────────────────────────────────────────

export async function storeReceipt(receipt: VoteReceipt): Promise<void> {
  const key   = `receipt:${receipt.bill_id}`;
  const bytes = utf8ToBytes(JSON.stringify(receipt));
  await dbSet(key, bytes);
}

export async function loadReceipt(billId: string): Promise<VoteReceipt | undefined> {
  const bytes = await dbGet(`receipt:${billId}`);
  if (!bytes) return undefined;
  return JSON.parse(new TextDecoder().decode(bytes)) as VoteReceipt;
}

// ─── Identity reset ───────────────────────────────────────────────────────────

/**
 * Wipes the local identity.
 * WARNING: Irreversible unless user re-registers with their phone.
 */
export async function clearIdentity(): Promise<void> {
  await dbDelete("nullifier_root");
}

File: apps/api/crypto/nullifier.py
"""
Server-side validation for Tier 1 HMAC-Nullifier-Chain voting protocol.

Validates:
  - Registration payloads (identity_commitment)
  - Vote payloads (ephemeral key + Ed25519 signature + vote_nullifier)
  - Issues VoteReceipts (server-signed acknowledgement)

Does NOT contain key derivation (that's client-side only).
"""
from __future__ import annotations

import struct
import time
from dataclasses import dataclass
from typing import Set

from nacl.signing import SigningKey, VerifyKey
from nacl.exceptions import BadSignatureError

# ── Protocol constants ────────────────────────────────────────────────────────

PROTO_VERSION = "ekklesia:v1"
TIMESTAMP_WINDOW_MS = 300_000  # 5 minutes
VALID_CHOICES = {"YES", "NO", "ABSTAIN"}


# ── Data classes ──────────────────────────────────────────────────────────────

@dataclass
class RegistrationPayload:
    identity_commitment: str  # 64-char hex (32 bytes)
    phone_region: str         # e.g. "GR"
    version: str


@dataclass
class VotePayload:
    bill_id: str
    choice: str               # YES | NO | ABSTAIN
    pk_eph: str               # 64-char hex — ephemeral public key
    vote_nullifier: str       # 64-char hex
    linkage_tag: str          # 64-char hex
    signature: str            # 128-char hex — Ed25519 sig
    timestamp_ms: int
    version: str


@dataclass
class VoteReceipt:
    bill_id: str
    vote_nullifier: str
    server_signature: str     # 128-char hex
    server_pk: str            # 64-char hex
    server_timestamp_ms: int


@dataclass
class ValidationError:
    code: str
    message: str


# ── Helpers ───────────────────────────────────────────────────────────────────

def _is_hex(s: str, expected_len: int) -> bool:
    """Check if string is valid hex of expected length."""
    if len(s) != expected_len:
        return False
    try:
        bytes.fromhex(s)
        return True
    except ValueError:
        return False


# ── Canonical payload builder ─────────────────────────────────────────────────

def build_signed_payload(
    *,
    bill_id: str,
    choice: str,
    pk_eph: bytes,
    vote_nullifier: bytes,
    linkage_tag: bytes,
    timestamp_ms: int,
) -> bytes:
    """
    Build the canonical byte sequence that the client signs.
    Must match the TypeScript client implementation exactly.

    Format: bill_id_utf8 | choice_utf8 | pk_eph(32) | vote_nullifier(32) | linkage_tag(32) | timestamp_ms(big-endian u64)
    """
    return (
        bill_id.encode("utf-8")
        + choice.encode("utf-8")
        + pk_eph
        + vote_nullifier
        + linkage_tag
        + struct.pack(">Q", timestamp_ms)
    )


# ── Registration validation ──────────────────────────────────────────────────

def validate_registration(payload: RegistrationPayload) -> ValidationError | None:
    """Validate a registration payload. Returns None on success."""
    if payload.version != PROTO_VERSION:
        return ValidationError("VERSION_MISMATCH", f"Expected {PROTO_VERSION}")

    if not _is_hex(payload.identity_commitment, 64):
        return ValidationError("INVALID_COMMITMENT", "identity_commitment must be 64-char hex")

    if not payload.phone_region or len(payload.phone_region) < 2:
        return ValidationError("INVALID_REGION", "phone_region required (e.g. 'GR')")

    return None


# ── Vote validation ──────────────────────────────────────────────────────────

def validate_vote(
    payload: VotePayload,
    used_nullifiers: Set[str],
    *,
    now_ms: int | None = None,
) -> ValidationError | None:
    """
    Validate a vote payload. Returns None on success, ValidationError on failure.

    Checks: version, choice, bill_id, hex formats, timestamp window,
    duplicate nullifier, Ed25519 signature.
    """
    if now_ms is None:
        now_ms = int(time.time() * 1000)

    # Version
    if payload.version != PROTO_VERSION:
        return ValidationError("VERSION_MISMATCH", f"Expected {PROTO_VERSION}")

    # Bill ID
    if not payload.bill_id:
        return ValidationError("INVALID_BILL_ID", "bill_id required")

    # Choice
    if payload.choice not in VALID_CHOICES:
        return ValidationError("INVALID_CHOICE", f"Must be one of {VALID_CHOICES}")

    # Hex format checks
    for field, name in [
        (payload.pk_eph, "pk_eph"),
        (payload.vote_nullifier, "vote_nullifier"),
        (payload.linkage_tag, "linkage_tag"),
    ]:
        if not _is_hex(field, 64):
            return ValidationError("INVALID_HEX", f"{name} must be 64-char hex")

    if not _is_hex(payload.signature, 128):
        return ValidationError("INVALID_SIGNATURE_FORMAT", "signature must be 128-char hex")

    # Timestamp window
    delta = abs(now_ms - payload.timestamp_ms)
    if delta > TIMESTAMP_WINDOW_MS:
        return ValidationError("TIMESTAMP_EXPIRED", f"Timestamp delta {delta}ms exceeds window")

    # Duplicate nullifier
    if payload.vote_nullifier in used_nullifiers:
        return ValidationError("DUPLICATE_VOTE", "vote_nullifier already used")

    # Ed25519 signature verification
    try:
        vk = VerifyKey(bytes.fromhex(payload.pk_eph))
        signed_bytes = build_signed_payload(
            bill_id=payload.bill_id,
            choice=payload.choice,
            pk_eph=bytes.fromhex(payload.pk_eph),
            vote_nullifier=bytes.fromhex(payload.vote_nullifier),
            linkage_tag=bytes.fromhex(payload.linkage_tag),
            timestamp_ms=payload.timestamp_ms,
        )
        vk.verify(signed_bytes, bytes.fromhex(payload.signature))
    except BadSignatureError:
        return ValidationError("INVALID_SIGNATURE", "Ed25519 signature verification failed")
    except Exception as e:
        return ValidationError("CRYPTO_ERROR", f"Crypto error: {type(e).__name__}")

    return None


# ── VoteReceipt issuance ─────────────────────────────────────────────────────

def issue_receipt(
    bill_id: str,
    vote_nullifier: str,
    server_sk: SigningKey,
) -> VoteReceipt:
    """
    Issue a server-signed VoteReceipt (ADR-008).
    Proves the server acknowledged this vote at this time.
    """
    ts = int(time.time() * 1000)
    receipt_bytes = (
        bill_id.encode("utf-8")
        + bytes.fromhex(vote_nullifier)
        + struct.pack(">Q", ts)
    )
    signed = server_sk.sign(receipt_bytes)

    return VoteReceipt(
        bill_id=bill_id,
        vote_nullifier=vote_nullifier,
        server_signature=bytes(signed.signature).hex(),
        server_pk=bytes(server_sk.verify_key).hex(),
        server_timestamp_ms=ts,
    )
